libkernelflinger/ivshmem.c

```diff
@@ -0,0 +1,413 @@
+#define PCI_CONFIG_CAP_PTR_OFFSET \
+    offsetof(struct pci_type0_config, capabilities_pointer)
+
+/*
+ * Memory Space bit in Command Register.
+ * Control a device's response to Memory Space access.
+ */
+#define CMD_MEM_SPACE_BIT_POSITION 1
+
+/**
+ * union pci_config_address - PCI configuration address
+ * @bits:           PCI configuration address in bits
+ * @bits.reg:       Register ID of PCI device
+ * @bits.function:  Function ID of PCI device
+ * @bits.device:    Device ID of PCI device
+ * @bits.bus:       Bus ID of PCI device
+ * @bits.reserved:  Reserved fields
+ * @bits.enable:    Enable bit
+ * @uint32:         32-bit value of union
+ */
+typedef union {
+	struct {
+		UINT32 reg:8;
+		UINT32 function:3;
+		UINT32 device:5;
+		UINT32 bus:8;
+		UINT32 reserved:7;
+		UINT32 enable:1;
+	} __attribute__((packed))  bits;
+	UINT32 uint32;
+} __attribute__((packed))  pci_config_address_t;
+
+struct ivshmem_device g_ivshmem_dev;
+
+static UINT8 hw_read_port_8(UINT16 port)
+{
+	UINT8 val8;
+
+	__asm__ __volatile__ (
+		"in %1, %0"
+		: "=a" (val8)
+		: "d" (port)
+	);
+
+	return val8;
+}
+
+static UINT16 hw_read_port_16(UINT16 port)
+{
+    UINT16 val16;
+
+    __asm__ __volatile__ (
+        "in %1, %0"
+        : "=a" (val16)
+        : "d" (port)
+    );
+
+    return val16;
+}
+
+static UINT32 hw_read_port_32(UINT16 port)
+{
+    UINT32 val32;
+
+    __asm__ __volatile__ (
+        "in %1, %0"
+        : "=a" (val32)
+        : "d" (port)
+    );
+
+    return val32;
+}
+
+static void hw_write_port_8(UINT16 port, UINT8 val8)
+{
+    __asm__ __volatile__ (
+        "out %1, %0"
+        :
+        : "d" (port), "a" (val8)
+    );
+}
+
+static void hw_write_port_16(UINT16 port, UINT16 val16)
+{
+    __asm__ __volatile__ (
+        "out %1, %0"
+        :
+        : "d" (port), "a" (val16)
+    );
+}
+
+static void hw_write_port_32(UINT16 port, UINT32 val32)
+{
+    __asm__ __volatile__ (
+        "out %1, %0"
+        :
+        : "d" (port), "a" (val32)
+    );
+}
+
+static UINT8 pci_read8(UINT8 bus, UINT8 device, UINT8 function, UINT8 reg)
+{
+    pci_config_address_t addr;
+
+    addr.uint32 = 0;
+    addr.bits.bus = bus;
+    addr.bits.device = device;
+    addr.bits.function = function;
+    addr.bits.reg = reg;
+    addr.bits.enable = 1;
+
+    hw_write_port_32(PCI_CONFIG_ADDRESS_REGISTER, addr.uint32 & ~0x3);
+    return hw_read_port_8(PCI_CONFIG_DATA_REGISTER | (addr.uint32 & 0x3));
+}
+
+static UINT16 pci_read16(UINT8 bus, UINT8 device, UINT8 function, UINT8 reg)
+{
+    pci_config_address_t addr;
+
+    addr.uint32 = 0;
+    addr.bits.bus = bus;
+    addr.bits.device = device;
+    addr.bits.function = function;
+    addr.bits.reg = reg;
+    addr.bits.enable = 1;
+
+    hw_write_port_32(PCI_CONFIG_ADDRESS_REGISTER, addr.uint32 & ~0x3);
+    return hw_read_port_16(PCI_CONFIG_DATA_REGISTER | (addr.uint32 & 0x3));
+}
+
+static UINT32 pci_read32(UINT8 bus, UINT8 device, UINT8 function, UINT8 reg)
```

Comment:
for these functions, including IO and PCI, does KF already has such implementations? @Huang, Yang please double check it.

Reply:
Yes, it has. But all are static functions.